internal/common/tree.go

```diff
@@ -0,0 +1,201 @@
+func (n *DNSTreeNode) RemoveNode(deleteNode *DNSTreeNode) {
+	// store indexes of dead branches
+	var deadBranches []int
+	var deadDataSets []string
+
+	if deleteNode == nil {
+		return
+	}
+
+	for i, node := range n.Children {
+		if node.Name == deleteNode.Name {
+			n.Children = append(n.Children[:i], n.Children[i+1:]...)
+
+			// we removed child, but need to clean up data sets
+			for j, dataset := range n.DataSets {
+				index := slices.Index(dataset.Targets, deleteNode.Name)
+				if index >= 0 {
+					n.DataSets[j].Targets = append(dataset.Targets[:index], dataset.Targets[index+1:]...)
+				}
+			}
+			return
+		}
+
+		// no children matched, try on each child
+		node.RemoveNode(deleteNode)
+
+		// the removed node was the only child - we have a dead branch
+		// children are nil on leaf node
+		// not checking for it will nuke the whole tree as leafs will be considered dead branches
+		if node.Children != nil && len(node.Children) == 0 {
+			// store the index. indexes are in ascending order
+			deadBranches = append(deadBranches, i)
+
+			// we can't rely on indexes for data sets, so store node name
+			deadDataSets = append(deadDataSets, node.Name)
+		}
+	}
+
+	// prune dead branches separately from the main for loop.
+	// doing it inside will shift indexes the for loop is iterating through
+	for count, deadBranchIndex := range deadBranches {
+		// after the first removal, all subsequent deadBranchIndexes will de one to high,
+		// but since we have them ascending, we can use count as modifier
+		n.Children = append(n.Children[:deadBranchIndex-count], n.Children[deadBranchIndex-count+1:]...)
+	}
+
+	var healthyDataSets []DNSTreeNodeData
+	// clean up data nodes from dead branches
+	for _, dataSet := range n.DataSets {
+
+		// we are dealing with CNAMES only here.
+		// the A record is already removed from datasets
+		if !slices.Contains(deadDataSets, dataSet.Targets[0]) {
+			healthyDataSets = append(healthyDataSets, dataSet)
+		}
+	}
+	n.DataSets = healthyDataSets
+
+}
+
+// GetLeafsTargets returns IP or CNAME of the leafs of a tree.
+// alternatively, it can populate the passed in array with pointers to targets
+func GetLeafsTargets(node *DNSTreeNode, targets *[]string) *[]string {
+	if node == nil || targets == nil {
+		return &[]string{}
+	}
+
+	// no children means this is pointing to an IP or a host outside the DNS Record
+	if len(node.Children) == 0 {
+		*targets = append(*targets, node.Name)
+		return nil
+	}
+	for _, child := range node.Children {
+		GetLeafsTargets(child, targets)
+	}
+	return targets
+}
+
+// ToEndpoints transforms a tree into an array of endpoints.
+// The array could be returned or passed in to be populated
+func ToEndpoints(node *DNSTreeNode, endpoints *[]*endpoint.Endpoint) *[]*endpoint.Endpoint {
+	if node == nil || endpoints == nil {
+		return &[]*endpoint.Endpoint{}
+	}
+
+	// no children means this is pointing to an IP or a host outside of the DNS Record
+	if len(node.Children) == 0 {
```

Comment:
I wonder would it read a little better if this check was done in a function name something like ```isFinalDNSEntry(node *DNSTreeNode)bool```